### PR TITLE
Machine Config Server implementation

### PR DESCRIPTION
**Pull request by vikramsk**
Support for serving the machine config in two modes:

- Cluster: when the server fetches the machine config from within the cluster.
- Bootstrap: when the server fetches the machine config from the file system of the machine its running on.

### CODE REVIEW

File: pkg/server/bootstrap_server.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+	// machineFileSystem defines the filesystem to be used
+	// for storing the files created by the server.
+	// eg: kubeconfig, node-annotations.
+	machineFileSystem string
+}
+
+// NewBootstrapServer initializes a new Bootstrap server that implements
+// the Server interface.
+func NewBootstrapServer(sfs, dir, mfs, kubeConf string) (Server, error) {
```
Comment (ashcrow) [inline pkg/server/bootstrap_server.go]: Since this does not produce an error, why have it as a return?

File: pkg/server/bootstrap_server.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	currConf := mp.Status.CurrentMachineConfig
+
+	// 2. Read the Machine Config object.
+	fileName = path.Join(bsc.serverBaseDir, cr.machinePool, "configs", currConf+".yaml")
+	data, err = ioutil.ReadFile(fileName)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("server: could not unmarshal file %s, err: %v", fileName, err)
```
Comment (ashcrow) [inline pkg/server/bootstrap_server.go]: I believe this is pre unmarshaling for section 2. `"server: could not read file %s, err: %v"` may be more appropriate.

File: pkg/server/bootstrap_server.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+// based on the pool request.
+// It returns nil for conf, error if the config isn't found. It returns a formatted
+// error if any other error is encountered during its operations.
+//
+// The method does the following:
+//
+// 1. Read the machine config pool by using the following path template:
+// 		"/<serverFileSystem>/<serverBaseDir>/<machineConfiPoolName>.yaml"
+//
+// 2. Read the currentConfig field from the Status and read the config file
```
Comment (abhinavdahiya) [inline pkg/server/bootstrap_server.go]: The ignition server should user the machineconfig specified in the machineconfigpool's .status.currentMachineConfig.
That means this dir structure is not required. just read the <machineconfig-name.yaml> ?

File: pkg/server/cluster_server.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("server: could not template etcd. err: %v", err)
+	}
+
+	err = appendMachineAnnotations(&mc.Spec.Config, csc.machineFileSystem, MachineAnnotationsPath, currConf)
+	if err != nil {
+		return nil, err
+	}
+
+	// append KubeConfig to Ignition.
```
Comment (abhinavdahiya) [inline pkg/server/cluster_server.go]: https://github.com/coreos-inc/tectonic-operators/blob/fe76a72d4ea3dfd6d9f7177ed751e4d8e379d83e/controller/node/pkg/tnc/tnc.go#L565

File: pkg/server/api.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		etcdIndex:   r.URL.Query().Get(paramEtcdIndex),
+	}
+
+	conf, err := sh.server.GetConfig(cr)
+
+	if conf == nil && err == nil {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	if err != nil {
```
Comment (yifan-gu) [inline pkg/server/api.go]: move this in front of the 404 check?

File: pkg/server/bootstrap_server.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+// GetConfig fetches the machine config(type - Ignition) from the bootstrap server,
+// based on the pool request.
+// It returns nil for conf, error if the config isn't found. It returns a formatted
+// error if any other error is encountered during its operations.
+//
+// The method does the following:
+//
+// 1. Read the machine config pool by using the following path template:
+// 		"/<serverFileSystem>/<serverBaseDir>/<machineConfiPoolName>.yaml"
```
Comment (yifan-gu) [inline pkg/server/bootstrap_server.go]: s/machineConfiPoolName/machineConfigPoolName

**Pull request merged**